### PR TITLE
Ranking Manager

### PR DESCRIPTION
**Pull request by dekelcohen**
New Ranking manager that sorts resulting tags in descending order according to combined score:

1) Scope Filter - filters out tags defined outside of scope (optional if tag generator supports it)
2) Same file and "this". ranking

3) Object Member Expression File Ranking
4) Rank by Definition Types: Rank Higher tags definitions with types matching to the GotoDef <reference> type

Some of the ranking / filtering require optional language syntax settings (see JS,Java and python examples in setting file). If not specified - the specific ranking method is skipped (or not optimal).

Tested with very large JavaScript project (using https://github.com/dekelcohen/estr JS tag generator) and with ctags for python, Java and Ruby projects (brief testing for non-JS languages). 

I believe that with this Ranking methods in place, we are making big steps with GotoDefinition ranking quality towards WebStorm 10.1. As an example,  with my large JS project, I often get better than WebStorm results.

### CODE REVIEW

**Comment by stephenfin** [comment]
This looks like some good work. I'll try my best to get this reviewed before the weekend!

**Comment by dekelcohen** [comment]
Hi Stephen,
Any news regarding the RankManager merge ?
I'm using it with different projects and languages, and didn't observe issues.

**Comment by stephenfin** [comment]
@dekelcohen I'm so sorry about not getting around to this - I no longer use this plugin so it's tough to find time/motivation to review PRs and fix bugs. *However*, I will get around to reviewing this tonight and if it's good I'll merge and generate a new release.

File: ranking/parse.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+class Parser:
+    """
+    Parses tag references and tag definitions. Used for ranking
+    """
+    @staticmethod
+    def extract_member_exp(line_to_symbol,source):
+        """
+        Extract receiver object e.g. receiver.mtd() 
+        Strip away brackets and operators.
+        TODO:HIGH: Add base lang defs + Python/Ruby/C++/Java/C#/PHP overrides (should be very similar)
```
Comment (stephenfin) [inline ranking/parse.py]: Is this TODO still valid. What are your plans for resolving this? What happens if we use a language that doesn't have a spec included in the settings?

**Comment by stephenfin** [comment]
One big comment. In addition, could you run through this and wrap any lines >80 chars - it makes it very hard to compare side-by-side

**Comment by dekelcohen** [comment]
Hi,
The languages defs are optional, so it will keep working with all langs.
I will however add more definitions soon because  most of the ranking
methods requires them + most curly braces langs have similar defs.
Example: Rank higher definition in the currently opened file - doesn't
require any per-lang definitions, but if we want to to take 'this'
expression into account, lang def for 'me/self/this' are required. Note: If
the defs are not complete or accurate it shouldn't cause a regression
(compared to no ranking).

I will also format >80 lines.
Thanks a lot.

**Comment by stephenfin** [comment]
Good stuff. I've tested it (albeit briefly) and it works rather well. I think I'm happy to merge once that little reformat is done (just run it through `autopep8` if you want).

**Comment by dekelcohen** [comment]
I pushed f9ffaeb - autopep8. Do I need to create a new Pull Request ?

**Comment by stephenfin** [comment]
Nope. I'll merge that this week. Thanks!

**Pull request merged**